blueocean-commons/src/main/java/io/jenkins/blueocean/commons/BlueoceanUrl.java

```diff
@@ -0,0 +1,285 @@
+     * @return A {@link BlueoceanUrl} instance iff the URL is a Blue Ocean UI URL, otherwise {@code null}.
+     */
+    public static @CheckForNull BlueoceanUrl parse(@Nonnull String url) {
+        Iterator<String> urlTokens = extractTokens(url);
+
+        if (urlTokens.hasNext()) {
+            if (urlTokens.next().equalsIgnoreCase("blue")) {
+                BlueoceanUrl blueoceanUrl = new BlueoceanUrl();
+
+                if (urlTokens.hasNext()) {
+                    String next = urlTokens.next();
+
+                    if (next.equalsIgnoreCase("pipelines")) {
+                        // i.e. /blue/pipelines/
+                        blueoceanUrl.addPart(UrlPart.DASHBOARD_PIPELINES, next);
+                    } else if (next.equalsIgnoreCase("organizations")) {
+                        // i.e. /blue/organizations/...
+                        if (urlTokens.hasNext()) {
+                            // e.g. /blue/organizations/jenkins/...
+                            blueoceanUrl.addPart(UrlPart.ORGANIZATION, urlTokens.next());
+                            if (urlTokens.hasNext()) {
+                                // e.g. /blue/organizations/jenkins/f1%2Ff3%20with%20spaces%2Ff3%20pipeline/...
+                                blueoceanUrl.addPart(UrlPart.PIPELINE, urlDecode(urlTokens.next()));
+                                if (urlTokens.hasNext()) {
+                                    next = urlTokens.next();
+                                    if (next.equalsIgnoreCase("detail")) {
+                                        // e.g. /blue/organizations/jenkins/f1%2Ff3%20with%20spaces%2Ff3%20pipeline/detail/...
+                                        blueoceanUrl.addPart(UrlPart.PIPELINE_RUN_DETAIL, next);
+                                        if (urlTokens.hasNext()) {
+                                            // e.g. /blue/organizations/jenkins/f1%2Ff3%20with%20spaces%2Ff3%20pipeline/detail/magic-branch-X/...
+                                            blueoceanUrl.addPart(UrlPart.BRANCH, urlDecode(urlTokens.next()));
+                                            if (urlTokens.hasNext()) {
+                                                // e.g. /blue/organizations/jenkins/f1%2Ff3%20with%20spaces%2Ff3%20pipeline/detail/magic-branch-X/55/...
+                                                blueoceanUrl.addPart(UrlPart.PIPELINE_RUN_DETAIL_ID, urlDecode(urlTokens.next()));
+                                                if (urlTokens.hasNext()) {
+                                                    next = urlTokens.next();
+                                                    if (PIPELINE_RUN_DETAIL_TABS.contains(next.toLowerCase())) {
+                                                        // e.g. /blue/organizations/jenkins/f1%2Ff3%20with%20spaces%2Ff3%20pipeline/detail/magic-branch-X/55/pipeline
+                                                        blueoceanUrl.addPart(UrlPart.PIPELINE_RUN_DETAIL_TAB, next.toLowerCase());
```

Comment:
@tfennelly I am ashamed to admit that this is quite easy to read like this! - I could see quickly that it was deliving into the URL hierarchy - there is probably a list/map way of doing it that keeps that clear hierarchy though.